spec/app/ics-029-fee-payment/README.md

```diff
@@ -0,0 +1,433 @@
+  denom: string,
+  amount: uint256,
+}
+```
+
+### IBC Module Wrapper
+
+The fee middleware will implement its own ICS-26 callbacks that wrap the application-specific module callbacks as well as the ICS-4 handler functions called by the underlying application. This fee middleware will ensure that the counterparty module supports incentivization and will implement all fee-specific logic. It will then pass on the request to the embedded application module for further callback processing.
+
+In this way, custom fee-handling logic can be hooked up to the IBC packet flow logic without placing the code in the ICS-4 handlers or the application code. This is valuable since the ICS-4 handlers should only be concerned with correctness of core IBC (transport, authentication, and ordering), and the application handlers should not be handling fee logic that is universal amongst all other incentivized applications. In fact, a given application module should be able to be hooked up to any fee module with no further changes to the application itself.
+
+#### Fee Protocol Negotiation
+
+The fee middleware will negotiate its fee protocol version with the counterparty module by prepending its own version to the application version. 
+
+Channel Version: `fee_v{fee_protocol_version}:{application_version}`
+
+Ex: `fee_v1:ics20-1`
+
+The fee middleware's handshake callbacks ensure that both modules agree on compatible fee protocol version(s), and then pass the application-specific version string to the embedded application's handshake callbacks.
+
+#### Handshake Callbacks
+
+```typescript
+function onChanOpenInit(
+  order: ChannelOrder,
+  connectionHops: [Identifier],
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier,
+  counterpartyPortIdentifier: Identifier,
+  counterpartyChannelIdentifier: Identifier,
+  version: string) {
+    // remove the prefix and pass the app-specific version to app callback.
+    // otherwise, pass version directly to app callback.
+    feeVersion, appVersion = splitFeeVersion(version)
+    // check that feeVersion is supported
+    if !isSupported(feeVersion) {
+        return error
+    }
+    app.OnChanOpenInit(
+        order,
+        connectionHops,
+        portIdentifier,
+        channelIdentifier,
+        counterpartyPortIdentifier,
+        counterpartyChannelIdentifier,
+        appVersion,
+    )
+}
+
+function OnChanOpenTry(
+  order: ChannelOrder,
+  connectionHops: [Identifier],
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier,
+  counterpartyPortIdentifier: Identifier,
+  counterpartyChannelIdentifier: Identifier,
+  version: string,
+  counterpartyVersion: string) {
+      // select mutually compatible fee version
+      cpFeeVersion, cpAppVersion = splitFeeVersion(counterpartyVersion)
+      feeVersion, appVersion = splitFeeVersion(version)
+      if !isCompatible(cpFeeVersion, feeVersion) {
+          return error
+      }
+      selectFeeVersion(cpFeeVersion, feeVersion)
```

Comment:
What is supposed to happen here?

Reply:
This is a middleware-specific function to negotiate the fee versions on each end of the channel. Given the counterparty fee version and the fee version, a final version for both modules is selected.